File: Week_09/java/src/_459_RepeatedSubstringPattern.java
/**
 * 第一遍：2020/08/24周一 ✅
 * 第二遍：2020/08/24周二 ✅
 * 第三遍：2020/08/24周三 ✅
 * 第四遍：2020/06/13周六
 */
class _459_RepeatedSubstringPattern {
    //时间：O(n^2) 空间：O(1)
    public boolean repeatedSubstringPattern(String s) {
        if (s == null) return false;
        int n = s.length();
        for (int i = 1; i <= (n >> 1); i++) {
            if ((n % i == 0)) {
                int j = i;
                while (j < n && s.charAt(j) == s.charAt(j - i)) {
                    j++;
                }
                if (j == n) {
                    return true;
                }
            }
        }
        return false;
    }

    class SolutionKMP {
        // 时间O(n) 空间O(n)
        public boolean repeatedSubstringPattern(String s) {
            if (s == null) return false;
            int m = s.length();
            int[] lps = new int[m];
            int j = 0;
            int i = 1;
            while (i < m) {
                if (s.charAt(i) == s.charAt(j)) {
                    lps[i++] = ++j;
                } else if (j != 0) {
                    j = lps[j - 1];
                } else {
                    lps[i++] = 0;
                }
            }

            return j != 0 && m % (m - j) == 0;
        }

    }
}

File: Week_09/kotlin/src/_459_RepeatedSubstringPattern.kt
fun repeatedSubstringPattern(s: String): Boolean {
    val n = s.length
    for (i in 1..(n ushr 1)) {
        if (n % i == 0) {
            var j = i
            while (j < n) {
                if (s[j] != s[j - i]) {
                    break
                }
                j++
            }
            if (j == n) {
                return true
            }
        }
    }
    return false
}

class _459_KMP_ {
    fun repeatedSubstringPattern(s: String): Boolean {
        val n = s.length
        val lps = IntArray(n)
        var i = 1
        var j = 0
        while (i < n) {
            when {
                s[j] == s[i] -> lps[i++] = ++j
                j != 0 -> j = lps[j - 1]
                else -> i++
            }
        }
        return j != 0 && n % (n - j) == 0
    }
}